service/src/test/java/bio/terra/workspace/app/configuration/external/controller/FolderApiControllerTest.java

```diff
@@ -0,0 +1,310 @@
+        .andExpect(status().is(HttpStatus.SC_NOT_FOUND));
+  }
+
+  @Test
+  public void listFolders_listAllFoldersInAWorkspace() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+    var description = String.format("This is folder %s", displayName);
+
+    ApiFolderDescription firstFolder =
+        createFolder(workspaceId, displayName, description, /*parentFolderId=*/ null);
+    ApiFolderDescription secondFolder =
+        createFolder(
+            workspaceId, displayName, description, /*parentFolderId=*/ firstFolder.getId());
+
+    String foldersGetResponse =
+        mockMvc
+            .perform(addAuth(get(String.format(FOLDERS_V1_PATH_FORMAT, workspaceId)), USER_REQUEST))
+            .andExpect(status().is(HttpStatus.SC_OK))
+            .andReturn()
+            .getResponse()
+            .getContentAsString();
+    var retrievedFolders =
+        objectMapper.readValue(foldersGetResponse, ApiFolderDescriptionsList.class);
+
+    var expectedFolders =
+        new ApiFolderDescriptionsList().addFoldersItem(firstFolder).addFoldersItem(secondFolder);
+    assertEquals(expectedFolders, retrievedFolders);
+  }
+
+  @Test
+  public void updateFolders_folderUpdated() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+    var description = String.format("This is folder %s", displayName);
+
+    ApiFolderDescription firstFolder =
+        createFolder(workspaceId, displayName, description, /*parentFolderId=*/ null);
+    ApiFolderDescription secondFolder =
+        createFolder(
+            workspaceId, displayName, description, /*parentFolderId=*/ firstFolder.getId());
+
+    var newDisplayName = "sofoo";
+    var newDescription = "This is a very foo folder";
+    String serializedUpdateResponse =
+        mockMvc
+            .perform(
+                addAuth(
+                    patch(String.format(FOLDER_V1_PATH_FORMAT, workspaceId, secondFolder.getId()))
+                        .contentType(MediaType.APPLICATION_JSON_VALUE)
+                        .accept(MediaType.APPLICATION_JSON)
+                        .characterEncoding("UTF-8")
+                        .content(
+                            getUpdateRequestInJson(newDisplayName, newDescription, null, true)),
+                    USER_REQUEST))
+            .andExpect(status().is(HttpStatus.SC_OK))
+            .andReturn()
+            .getResponse()
+            .getContentAsString();
+    var updatedFolderDescription =
+        objectMapper.readValue(serializedUpdateResponse, ApiFolderDescription.class);
+
+    assertEquals(newDisplayName, updatedFolderDescription.getDisplayName());
+    assertEquals(newDescription, updatedFolderDescription.getDescription());
+    assertNull(updatedFolderDescription.getParentFolderId());
+  }
+
+  @Test
+  public void updateFolders_invalidParentFolder_throws404() throws Exception {
```

Comment:
nit what exactly does invalid mean, does parent folder not exist? Can you be clear

updateFolders_parentFolderDoesNotExist_throws404